broker-core/src/test/java/io/zeebe/broker/job/JobStateControllerTest.java

```diff
@@ -0,0 +1,263 @@
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldDeleteJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.create(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+
+    // when
+    stateController.delete(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isFalse();
+    assertThat(State.values()).noneMatch(state -> stateController.isInState(key, state));
+    assertThat(stateController.getJob(key)).isNull();
+  }
+
+  @Test
+  public void shouldFailJobWithRetriesLeft() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord().setRetries(1);
+
+    // when
+    stateController.fail(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldFailJobWithNoRetriesLeft() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord().setRetries(0);
+
+    // when
+    stateController.fail(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.FAILED);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldResolveJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.fail(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldListTimedOutEntriesInOrder() {
+    // given
+    stateController.activate(1, createJobRecord().setDeadline(1L));
+    stateController.activate(2, createJobRecord().setDeadline(256L));
+    stateController.activate(3, createJobRecord().setDeadline(65536L));
+    stateController.activate(4, createJobRecord().setDeadline(4294967296L));
+
+    // when
+    final List<Long> jobKeys = new ArrayList<>();
+    stateController.forEachTimedOutEntry(32768L, ((key, record, control) -> jobKeys.add(key)));
+
+    // then
+    assertThat(jobKeys).hasSize(2);
+    assertThat(jobKeys).containsExactly(1L, 2L);
+  }
+
+  @Test
+  public void shouldCheckExistenceCorrectly() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.create(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertThat(stateController.exists(key + 1)).isFalse();
+  }
+
+  @Test
+  public void shouldListActivatableJobsForTypeInOrder() {
+    // given
```

Comment:
What if no job with this type exist?

Reply:
Then the callback is never called. I guess then it's a semantics issue, where the stream processor will write a follow up event with intent `JobBatchIntent.ACTIVATED` even though nothing was activate (so it will contain an empty list). I don't think that's necessarily an error, but depends what API we want I guess.